zephyr: Add Seeed Studio's XIAO BLE NRF52840 SENSE board.

bikeNomad opened the pull request:
### Summary

This PR adds Zephyr support for Seeed Studio's XIAO BLE NRF52840 SENSE board.

It also provides a good example of a richer Zephyr port than the default, adding:
  - Frozen modules (including `asyncio`, `upysh`, `aioble`, `neopixel`, and `aiorepl`).
  - Enough MicroPython features to support using `aioble` (at least for the `temp_sensor.py` example)
  -  JSON, `random`, `re`, `struct`, etc.
  - Support for my Zephyr ADC code (see PR #17615)
  - Mounted LFS2 filesystem using the 2MiB flash mounted as `/flash`
 
You can build it in the usual way:
```
$ cd ports/zephyr
$ west build -b xiao_ble/nrf52840/sense . 
```
This will produce a `.uf2` file in `build/zephyr/zephyr.uf2` that can be copied into the XIAO-SENSE USB drive that appears when you double-click the `RESET` button.

### Testing

I have tested this on several of these boards, with and without my ADC (#17615) and `zensor.Sensor.attr_set()` (#17653) PRs.

I have run the `aioble` `temp_sensor.py` example but haven't done much more with the BLE.

### Trade-offs and Alternatives

The build size is as follows:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      428864 B       788 KB     53.15%
             RAM:      165908 B       256 KB     63.29%
        IDT_LIST:          0 GB        32 KB      0.00%
```

ned-pcs commented:
The additional features defined in `mpconfigport-xiao.h` add 12224 bytes of flash and 12 bytes of RAM size.

dpgeorge commented on ports/zephyr/boards/manifest_xiao_ble_nrf52840_sense.py:
```diff
@@ -0,0 +1,11 @@
+# This is an example frozen module manifest. Enable this by configuring
+# the Zephyr project and enabling the frozen modules config feature.
```
I think this comment can be removed, it's not really an example anymore for this board.

dpgeorge commented on ports/zephyr/boards/mpconfigport-xiao.h:
```diff
@@ -0,0 +1,51 @@
+
+#include "mpconfigport.h"
+
+#define MICROPY_PY_ASYNCIO (1)
```
Instead of defining all these options, it may be simpler to do:
```c
#define MICROPY_CONFIG_ROM_LEVEL (MICROPY_CONFIG_ROM_LEVEL_EXTRA_FEATURES)
```
Then you automatically get the new features at that level as they are added in the future, and the config matches other boards at this level.

ned-pcs replied:
Good idea, though it adds about 26kB to the flash size.

dpgeorge replied:
> it adds about 26kB to the flash size.

I think that's fine.  This part has enough flash, and it keeps the configuration simple.

dpgeorge commented on ports/zephyr/boards/xiao_ble_nrf52840_sense.conf:
```diff
@@ -0,0 +1,66 @@
+
+CONFIG_MICROPY_FROZEN_MODULES=y
+CONFIG_MICROPY_FROZEN_MANIFEST="boards/manifest_xiao_ble_nrf52840_sense.py"
+CONFIG_MICROPY_CONFIGFILE="boards/mpconfigport-xiao.h"
```
The `boards` directory is going to get rather messy with all these files.  I suggest making a subdir for these files:
```
boards/xiao_ble_nrf52840_sense/manifest.py
boards/xiao_ble_nrf52840_sense/mpconfigport.h
```
Note that the conf and overlay files need to remain where they are so west can find them.

bikeNomad commented:
@blafy asks:
>  Great ! Do you have a plan to implement a disable flag for the p25q16h ? This little flash is IMO a bad addition on this board, as it requires to be put into suspend mode to keep the nice low power of the NRF chip ...

Good question. Not using this flash means that instead of 2MiB filesystem we'd have only 32KiB, at least without partition modification.

The stock "Storage" partition (given the original overlay; see `nrf52840_partition_uf2_sdv7.dtsi` ) on the NRF chip is only 32kB in size (156kB is given to the SoftDevice that we don't use, and 48kB is saved for the UF2 bootloader). The way these boards are set up at the factory includes the UF2 bootloader and SoftDevice. Using MCUBoot and two app partitions (see `nrf52840_partition.dtsi`) still only leaves 32kB of storage space.

Although the newest Adafruit UF2 bootloader allows for having no SoftDevice (thus maybe letting us use the 156kiB space for `/flash` instead), I haven't tried this.

As for putting it into suspend mode, I would hope that the `nortic,qspi-nor` driver would do this when possible. I've seen current consumption in the 1-2mA range even without the new `lightsleep` code.

dpgeorge commented on ports/zephyr/boards/xiao_ble_nrf52840_sense.conf:
```diff
@@ -0,0 +1,67 @@
+
+# Enable MicroPython's VFS and mount a littlefs filesystem on boot
+CONFIG_MICROPY_VFS_LFS2=y
+CONFIG_MICROPY_VFS_FAT=y
```
I think these will be enabled by default from `prj.conf`, right?

If so, I suggest removing everything from here that's already the default in `prj.conf`, to keep this file minimal.  Otherwise in the future when people add other boards, they copy this file and there are more and more duplicated settings.

dpgeorge commented on ports/zephyr/boards/xiao_ble_nrf52840_sense/mpconfigport.h:
```diff
@@ -0,0 +1,32 @@
+#undef MICROPY_CONFIG_ROM_LEVEL
+#define MICROPY_CONFIG_ROM_LEVEL (MICROPY_CONFIG_ROM_LEVEL_EXTRA_FEATURES)
+
+#define MICROPY_PY_SYS_MODULES (1)
```
This is already the default (in `py/mpconfig.h`), so can be removed.

dpgeorge commented on ports/zephyr/boards/xiao_ble_nrf52840_sense/mpconfigport.h:
```diff
@@ -0,0 +1,32 @@
+#include "../mpconfigport.h"
+
+#undef MICROPY_CONFIG_ROM_LEVEL
+#define MICROPY_CONFIG_ROM_LEVEL (MICROPY_CONFIG_ROM_LEVEL_EXTRA_FEATURES)
```
If you set this before including `../mpconfigport.h` then it doesn't need to be undef'd.

dpgeorge commented on ports/zephyr/boards/xiao_ble_nrf52840_sense/manifest.py:
```diff
@@ -0,0 +1,8 @@
+# freeze("../modules")
```
This comment can be removed.

dpgeorge commented on ports/zephyr/boards/xiao_ble_nrf52840_sense/manifest.py:
```diff
@@ -0,0 +1,8 @@
+# freeze("../modules")
+include("$(MPY_DIR)/extmod/asyncio")
+
+# Require a micropython-lib module.
```
This comment is not really correct, it was copied from the example manifest.

You could remove it, or change it to something like "Require some useful micropython-lib modules."

jonnor commented:
I have tested this working on device. So that is a +1 for me. 

In general, I think this is a good target hardware for a standard build on a Zephyr board. For the following reasons:

- It is widely available (incl via distributors like Digikey etc) and affordable, with schematics etc available
- Easy to flash without special tools, since it implements UF2 over USB
- It covers NRF52, which is now in some areas better supported with Zephyr port than "nrf" port - notably Bluetooth Low Energy
- Has I2C devices on-board, enables testing that key module

jonnor commented:
@bikeNomad there are some checks failing for the commit formatting and code spelling. Can you check them out and resolve them?

bikeNomad commented:
> @bikeNomad there are some checks failing for the commit formatting and code spelling. Can you check them out and resolve them?
> 
> One of the failing check is for the name "Ned", which seems not-very-useful. But I have no idea how to make codespell ignore that, unfortunately. Or if that is critical wrt merging.

I'm working on getting the checks to pass. I found that I had to use all upper-case letters for `NED` because `codespell` is configured to ignore three-letter all-caps words.

bikeNomad commented:
I re-based this on the ADC PR (#17616) to simplify.

The other thing that I added was automatic filesystem creation in case of an unformatted flash storage area (which will be the case when applying to a new board). I did this in the same way that other ports do it: by adding a `_boot.py` module that first tries to mount an existing formatted filesystem, then formats and mounts if that fails.

Users of other zephyr ports have noticed the lack of automatic filesystem creation, as well.

I could separate this into a separate PR if desired.

bikeNomad commented:
> I dont think the bits on the FS should be accepted, the current code is fine for its purpose and this is just translating it into python for the most part, and disabling/breaking the feature for every other supported boards because you are not adding a generic manifest.py, which might break other boards in other ways anyway.

@VynDragon thanks for looking at the code. I'll remove the unnecessary lines in the config.

There is a generic `manifest.py` that freezes the `modules` directory (which currently contains only `_boot.py`). The current code doesn't bother to create a filesystem in the case where there isn't one already (it just mounts it if it's already formatted).

This lack of filesystem formatting has bitten myself and others (for instance: #16240). It will cause frustration for future users installing to boards for the first time.

I will, however, separate the FS code from the PR.

bikeNomad committed:
zephyr/boards: Add XIAO BLE NRF52840 SENSE board.

This commit adds Zephyr support for the XIAO BLE NRF52840 SENSE board from
Seeed Studio.

It also provides a good example of a richer Zephyr port than the default,
adding:
- Frozen modules (including asyncio, upysh, aioble and aiorepl).
- Enough MicroPython features to support using aioble (at least for the
  `temp_sensor.py` example).
- JSON, random, re, struct, etc.

Signed-off-by: Ned Konz <ned@metamagix.tech>
